visual-js/visual-playwright/src/api.ts

```diff
@@ -0,0 +1,325 @@
+
+export const getApi = (client = `visual-playwright/${clientVersion}`) => {
+  let api = globalThis.visualApi;
+  if (!api) {
+    api = createVisualApi(client);
+    globalThis.visualApi = api;
+  }
+  return api;
+};
+
+export const getBuildIdForCustomId = async (
+  customId: string,
+): Promise<string | null> => {
+  const build = await getApi().buildByCustomId(customId);
+  return build?.id ?? null;
+};
+
+export const createBuild = async () => {
+  const { buildName, project, branch, customId, defaultBranch } = getOpts();
+  const build = await getApi().createBuild({
+    name: buildName,
+    branch,
+    defaultBranch,
+    project,
+    customId,
+  });
+
+  if (!build) {
+    throw new Error(
+      "Failed to create Sauce Visual build. Please check that you've supplied valid credentials and selected an available region for your account and try again.",
+    );
+  }
+
+  console.info(
+    `View your in-progress build on the Sauce Labs dashboard: 
+  ${build.url}
+`,
+  );
+
+  return build.id;
+};
+
+export const finishBuild = async (id: string) => {
+  return await getApi().finishBuild({
+    uuid: id,
+  });
+};
+
+/**
+ * Wait for the result of the build. Retries gracefully until a build status !== running is met.
+ * Resolves with an exit code of 1 if unreviewed changes are found.
+ * @param buildId
+ */
+export const waitForBuildResult = async (buildId: string) => {
+  const buildInProgressError = new Error('Build execution in progress.');
+  const buildNotFoundError = new Error(
+    'Build has been deleted or you do not have access to view it.',
+  );
+
+  try {
+    const fetchBuild = async () => {
+      const build = await getApi().buildStatus(buildId);
+
+      if (!build) {
+        throw buildNotFoundError;
+      }
+
+      if (build.status !== BuildStatus.Running) {
+        return build;
+      }
+
+      throw buildInProgressError;
+    };
+
+    const build = await backOff(fetchBuild, {
+      maxDelay: 10000,
+      numOfAttempts: 20,
+      // continues to retry on true, on false will short-circuit and end retry early
+      retry: (e) => e !== buildNotFoundError,
+    });
+
+    console.info(
+      `Your Sauce Visual Build is ready for review:
+  ${build.url}
+`,
+    );
+
+    if (build.unapprovedCount > 0) {
+      process.exitCode = 1;
+    }
+  } catch (e) {
+    console.error(
+      `Failed to determine build completion status. This could be due to an abnormally long-running build, or an error in communication with the server. See the error message below for more details:
+      
+${e instanceof Error ? e.message : JSON.stringify(e)}
+`,
+    );
+
+    process.exitCode = 1;
+  }
+};
+
+/**
+ * Takes a snapshot of the current viewport and uploads it to Sauce Visual.
+ */
+export const sauceVisualCheck = async (
+  page: Page,
+  testInfo: TestInfo,
+  name: string,
+  options?: Partial<SauceVisualParams>,
+) =>
+  await takePlaywrightScreenshot(
+    page,
+    {
+      deviceName: testInfo.project.name,
```

Comment:
Is deviceName expected to contain project name ?

Reply:
Yeah, Playwright doesn't really expose 'devices', but instead [projects](https://playwright.dev/docs/test-projects) which can be used to create configurations for runs. They are labelled in the config so we would just apply those labels here since some may be emulated.